src/supervisor/mod.rs

```diff
@@ -64,16 +71,39 @@ pub struct Spec {
     pub meta: SpecMeta,
 }
 
+/// A value that gets returned when running a Supervisor record
+enum SupervisorResult {
+    // gets returned when the supervisor failed to start
+    ErrStart {
+        spec: Spec,
+        err: Arc<StartError>,
+    },
+    // gets returned when the supervisor failed to start, and the
+    // error is sent to a spawner thread via a start_notifier
+    ErrStartHandled,
+    // gets returned when a supervisor stopped running because of a termination
+    // that did not fail
+    OkTermination {
+        spec: Spec,
+    },
+    // gets returned when a supervisor stopped running because of a termination
+    // but it did find errors in the process
+    ErrTermination {
+        spec: Spec,
+        err: Arc<TerminationError>,
+    },
+}
+
 pub struct Supervisor {
     pub runtime_name: String,
-    pub children: Vec<Worker>,
-
     pub meta: SpecMeta,
+    join_handle: JoinHandle<SupervisorResult>,
+    termination_handle: AbortHandle,
 }
 
 // terminate_prev_siblings gets called when a child in a supervision tree fails
 // to start, and all it's previous siblings need to get terminated.
-async fn terminate_prev_siblings<I: DoubleEndedIterator<Item = worker::Spec>>(
+async fn terminate_prev_siblings<I: DoubleEndedIterator<Item = worker::Spec> + Send>(
```

Comment:
what is causing the need for this `Send` bound?

Reply:
Ah, for a second I needed it, but now is redundant, good find